File: dispatch/java/adamma/c4dhi/claid/Module/Channel.java
package adamma.c4dhi.claid.Module;

import java.util.function.Consumer;

import adamma.c4dhi.claid.Logger.Logger;

public class Channel<T>
{
    private final String channelId;
    private final ChannelAccessRights accessRights;
    private Module parent;

    private Publisher<T> publisher;
    private Subscriber<T> subscriber;
    boolean callbackRegistered = false;

    private boolean valid = false;

    private boolean canRead()
    {
        return this.accessRights == ChannelAccessRights.READ || this.accessRights == ChannelAccessRights.READ_WRITE;
    }

    private boolean canWrite()
    {
        return this.accessRights == ChannelAccessRights.WRITE || this.accessRights == ChannelAccessRights.READ_WRITE;
    }

    // Constructor for published Channels.
    public Channel(Module parent, final String channelId, final Publisher<T> publisher)
    {
        this.channelId = channelId;
        this.accessRights = ChannelAccessRights.WRITE;
        this.publisher = publisher;
        this.valid = true;
    }

    // Constructor for subscribed Chanenls.
    public Channel(Module parent, final String channelId, final Subscriber<T> subscriber)
    {
        this.channelId = channelId;
        this.accessRights = ChannelAccessRights.READ;
        this.subscriber = subscriber;
        this.callbackRegistered = subscriber != null;
        this.valid = true;
    }

    private Channel(final String channelId)
    {
        this.accessRights = ChannelAccessRights.NONE;
        this.channelId = channelId;
        this.parent = null;
        this.valid = false;
    }

    static<T> Channel<T> newInvalidChannel(final String channelId)
    {
        Channel<T> channel = new Channel<T>(channelId);
        return channel;
    }

    public void post(T data)
    {
        if(!canWrite())
        {
            String msg = "Module \"" + this.parent.getId() + "\"] tried to post data to channel \"" + this.channelId + "\", however\n" +    
                        "it did not publish this channel before.";
            parent.moduleError(msg);
            return;
        }
        this.publisher.post(data);
    }
}

File: dispatch/python/module/channel.py
from enum import Enum
from typing import Type, Generic

from module.publisher import Publisher
from module.subscriber import Subscriber
from module.channel_access_rights import ChannelAccessRights


class Channel():
    def __init__(self, channel_id: str = "", access_rights: ChannelAccessRights = ChannelAccessRights.NONE,
                 parent = None, publisher: Publisher = None, subscriber: Subscriber = None,
                 callback_registered: bool = False, valid: bool = False):
        self.channel_id = channel_id
        self.access_rights = access_rights
        self.parent = parent
        self.publisher = publisher
        self.subscriber = subscriber
        self.callback_registered = callback_registered
        self.valid = valid

    @classmethod
    def new_invalid_channel(cls, channel_id: str) -> 'Channel':
        return cls(channel_id=channel_id)

    @classmethod
    def published_channel(cls, parent, channel_id: str, publisher: Publisher) -> 'Channel':
        return cls(channel_id=channel_id, access_rights=ChannelAccessRights.WRITE, parent=parent, publisher=publisher, valid=True)

    @classmethod
    def subscribed_channel(cls, parent, channel_id: str, subscriber: Subscriber) -> 'Channel':
        return cls(channel_id=channel_id, access_rights=ChannelAccessRights.READ, parent=parent, subscriber=subscriber, callback_registered=subscriber is not None, valid=True)

    def can_read(self) -> bool:
        return self.access_rights in {ChannelAccessRights.READ, ChannelAccessRights.READ_WRITE}

    def can_write(self) -> bool:
        return self.access_rights in {ChannelAccessRights.WRITE, ChannelAccessRights.READ_WRITE}

    def post(self, data) -> None:
        if not self.can_write():
            msg = f"Tried to post data to channel \"{self.channel_id}\", however\nit did not publish this channel before."
            self.parent.module_error(msg)
            return

        self.publisher.post(data)
